pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixInstanceDataManager.java

```diff
@@ -311,6 +310,47 @@ private void reloadSegment(String tableNameWithType, SegmentMetadata segmentMeta
     }
   }
 
+  @Override
+  public void addOrReplaceSegment(String tableNameWithType, String segmentName, boolean forceDownload)
+      throws Exception {
+    LOGGER.info("Adding or replacing segment: {} for table: {} downloadIsForced: {}", segmentName, tableNameWithType,
+        forceDownload);
+
+    // Get updated table config and segment metadata from Zookeeper.
+    TableConfig tableConfig = ZKMetadataProvider.getTableConfig(_propertyStore, tableNameWithType);
+    Preconditions.checkNotNull(tableConfig);
+    SegmentZKMetadata zkMetadata =
+        ZKMetadataProvider.getSegmentZKMetadata(_propertyStore, tableNameWithType, segmentName);
+    Preconditions.checkNotNull(zkMetadata);
+
+    // This method might modify the file on disk. Use segment lock to prevent race condition
+    Lock segmentLock = SegmentLocks.getSegmentLock(tableNameWithType, segmentName);
+    try {
+      // Lock the segment to get its metadata, so that no other threads are modifying
+      // the disk files of this segment.
+      segmentLock.lock();
+      SegmentMetadata localMetadata = getSegmentMetadata(tableNameWithType, segmentName);
+
+      _tableDataManagerMap.computeIfAbsent(tableNameWithType, k -> createTableDataManager(k, tableConfig))
+          .addOrReplaceSegment(segmentName, new IndexLoadingConfig(_instanceDataManagerConfig, tableConfig),
+              localMetadata, zkMetadata, forceDownload);
+      LOGGER.info("Added or replaced segment: {} of table: {}", segmentName, tableNameWithType);
+    } finally {
+      segmentLock.unlock();
+    }
+  }
+
+  @Override
+  public void addOrReplaceAllSegments(String tableNameWithType, boolean forceDownload)
+      throws Exception {
+    LOGGER.info("Adding or replacing all segments in table: {} downloadIsForced: {}", tableNameWithType, forceDownload);
+    List<SegmentMetadata> segMds = getAllSegmentsMetadata(tableNameWithType);
+    for (SegmentMetadata segMd : segMds) {
+      addOrReplaceSegment(tableNameWithType, segMd.getName(), forceDownload);
```

Comment:
Add a helper method with `TableConfig` and `SegmentZKMetadata` as the arguments, and make `addOrReplaceSegment()` and `addOrReplaceAllSegments()` calling the helper method. We should avoid reading them for each segment for bulk replace

Reply:
after refactoring reloadSegment with forceDownload flag, there is no need for this method.